src/main/webapp/app/shared/util/reactive-filter.ts

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2021. The Hyve
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * See the file LICENSE in the root of this repository.
+ */
+
+import { BehaviorSubject, merge, Observable, Subject } from "rxjs";
+import { debounceTime, distinctUntilChanged, filter, shareReplay, startWith } from "rxjs/operators";
+import {
+  NgbCalendar,
+  NgbDate,
+  NgbDateParserFormatter,
+  NgbDateStruct
+} from "@ng-bootstrap/ng-bootstrap";
+
+export class ReactiveFilter<T> {
+  protected readonly _value$: BehaviorSubject<T>;
+  protected readonly _error$: BehaviorSubject<string>;
+  protected readonly trigger$: Subject<T>;
+
+  error$: Observable<string>;
+  rawValue$: Observable<T | null>;
+  value$: Observable<T | null>;
+
+  constructor(
+    options?: ReactiveFilterOptions<T>
+  ) {
+    if (!options) {
+      options = {};
+    }
+    this._error$ = new BehaviorSubject('');
+    this.error$ = this._error$.asObservable().pipe(distinctUntilChanged());
+
+    this._value$ = new BehaviorSubject<T>(null);
+    this.rawValue$ = this._value$.asObservable();
+    this.trigger$ = new Subject<T>();
+
+    let validValue = this._value$.pipe(
+        debounceTime(options.debounceTime || 200),
+        filter(v => {
+            const error = this.validate(v);
+            if (error) {
+                this._error$.next(error);
+                return false;
+            } else {
+                this._error$.next('');
+                return true;
+            }
+        }),
+    );
+    let mergedSignal = merge(validValue, this.trigger$).pipe(
+      startWith(options.initialValue || null as T | null),
+    );
+    if (options.mapResult) {
+      mergedSignal = options.mapResult(mergedSignal);
+    } else {
+      mergedSignal = mergedSignal.pipe(distinctUntilChanged());
+    }
+    this.value$ = mergedSignal.pipe(shareReplay(1));
+  }
+
+  next(value?: T){
+    this._value$.next(value);
+  }
+
+  clear() {
+    this._value$.next(null);
+    this._error$.next('');
+    this.trigger$.next(null);
+  }
+
+  complete() {
+    this._value$.complete();
+    this._error$.complete();
+    this.trigger$.complete();
+  }
```

Comment:
This method is not used. Shouldn't we call it the onDestroy method of subject.component.ts?

Reply:
My editor says the same, but it is called in SubjectComponent line 162:
```
        for (let filtersKey in this.filters) {
             this.filters[filtersKey].complete();
         }
```

Reply:
Yes, it's my mistake.